pallets/amm/src/lib.rs

```diff
@@ -158,16 +160,17 @@ pub mod pallet {
             minimum_amounts: (Balance, Balance),
         ) -> DispatchResultWithPostInfo {
             let who = ensure_signed(origin)?;
-            let (is_inverted, base_asset, quote_asset) = Self::get_upper_currency(pool.0, pool.1);
+            let (is_inverted, base_asset, quote_asset) =
+                Self::get_upper_currency(pool.0, pool.1).ok_or(Error::<T, I>::InvalidCurrencyId)?;
 
             let (base_amount, quote_amount) = match is_inverted {
                 true => (liquidity_amounts.1, liquidity_amounts.0),
                 false => (liquidity_amounts.0, liquidity_amounts.1),
             };
 
             Pools::<T, I>::try_mutate(
-                &base_asset,
-                &quote_asset,
+                base_asset.clone(),
+                quote_asset.clone(),
```

Comment:
Do you still need the `.clone()`? I think having the `Copy` trait should be enough to avoid this?